packages/protobuf/src/next/from-binary.ts

```diff
@@ -0,0 +1,274 @@
+// Copyright 2021-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { BinaryReadOptions } from "../binary-format.js";
+import type { DescField, DescMessage } from "../descriptor-set.js";
+import type { MessageShape } from "./types.js";
+import type { ReflectMessage } from "./reflect/reflect.js";
+import type { IBinaryReader } from "../binary-encoding.js";
+import {
+  LongType,
+  scalarZeroValue,
+  type ScalarValue,
+} from "./reflect/scalar.js";
+import { reflect } from "./reflect/reflect.js";
+import { create } from "./create.js";
+import { BinaryReader, WireType } from "../binary-encoding.js";
+import { ScalarType } from "./reflect/scalar.js";
+import {
+  FeatureSet_MessageEncoding,
+  FieldDescriptorProto_Type,
+} from "../google/protobuf/descriptor_pb.js";
+
+// Default options for parsing binary data.
+const readDefaults: Readonly<BinaryReadOptions> = {
+  readUnknownFields: true,
+  readerFactory: (bytes) => new BinaryReader(bytes),
+};
+
+function makeReadOptions(
+  options?: Partial<BinaryReadOptions>,
+): Readonly<BinaryReadOptions> {
+  return options ? { ...readDefaults, ...options } : readDefaults;
+}
+
+export function fromBinary<Desc extends DescMessage>(
+  desc: Desc,
+  bytes: Uint8Array,
+  options?: Partial<BinaryReadOptions>,
+): MessageShape<Desc> {
+  options = makeReadOptions(options);
+  return readMessage(
+    desc,
+    options.readerFactory!(bytes),
+    bytes.byteLength,
+    options as BinaryReadOptions,
+  );
+}
+
+function readMessage<Desc extends DescMessage>(
+  desc: Desc,
+  reader: IBinaryReader,
+  lengthOrEndTagFieldNo: number,
+  options: BinaryReadOptions,
+  delimitedMessageEncoding?: boolean,
+): MessageShape<Desc> {
```

Comment:
We need a better function signature. 

The signature is part of the v1 public API (see [here](https://github.com/bufbuild/protobuf-es/blob/c0d0553d24a5df462c3b2c775f9b86b50a22bbe8/packages/protobuf/src/binary-format.ts#L43-L61)), and is a bit odd because it needed to be backwards compatible.

Can you add a TODO?